web/federate.go

```diff
@@ -0,0 +1,111 @@
+package web
+
+import (
+	"io"
+	"net/http"
+
+	"bitbucket.org/ww/goautoneg"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/text"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage/local"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+	dto "github.com/prometheus/client_model/go"
+)
+
+type Federation struct {
+	Storage local.Storage
+}
+
+func (fed *Federation) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	req.ParseForm()
+
+	fps := map[clientmodel.Fingerprint]struct{}{}
+
+	for _, s := range req.Form["match[]"] {
+		matchers, err := promql.ParseMetricSelector(s)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		for _, fp := range fed.Storage.FingerprintsForLabelMatchers(matchers) {
+			fps[fp] = struct{}{}
+		}
+	}
+
+	enc, contentType := chooseEncoder(req)
+	w.Header().Set("Content-Type", contentType)
+
+	protMetric := &dto.Metric{
+		Label:   []*dto.LabelPair{},
+		Untyped: &dto.Untyped{},
+	}
+	protMetricFam := &dto.MetricFamily{
+		Metric: []*dto.Metric{protMetric},
+		Type:   dto.MetricType_UNTYPED.Enum(),
+	}
+
+	for fp := range fps {
+		met := fed.Storage.MetricForFingerprint(fp)
+		if met.Metric == nil {
+			continue
+		}
+		sp := fed.Storage.LastSamplePairForFingerprint(fp)
+		if sp == nil {
+			continue
+		}
+
+		// Reset label slice.
+		protMetric.Label = protMetric.Label[:0]
+
+		for ln, lv := range met.Metric {
+			if ln == clientmodel.MetricNameLabel {
+				protMetricFam.Name = (*string)(&ln)
+				continue
+			}
+			protMetric.Label = append(protMetric.Label, &dto.LabelPair{
+				Name:  (*string)(&ln),
+				Value: (*string)(&lv),
+			})
+		}
+		protMetric.TimestampMs = (*int64)(&sp.Timestamp)
+		protMetric.Untyped.Value = (*float64)(&sp.Value)
+
+		if _, err := enc(w, protMetricFam); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+type encoder func(w io.Writer, p *dto.MetricFamily) (int, error)
+
+func chooseEncoder(req *http.Request) (encoder, string) {
```

Comment:
I have no estimate. Should be a simple refactoring, no? ;-)

Intermediate buffering cannot be really avoided as we have a contract to sort metrics lexicographically. So after we have collected everything, we need to sort the protobuf. (In some cases, collection could be done smarter, but not always, see the metric family injection hook and the planned more flexible "prefix" collectors.)